traitsui/tests/editors/test_liststr_editor_selection.py

```diff
@@ -74,21 +92,539 @@ class ListStrEditorWithSelectedIndex(HasTraits):
+
+            # If a change in multi_selected involves an invalid value, nothing
+            # is changed
+            editor.multi_selected[0] = "four"
+            gui.process_events()
+
+            # FIXME issue enthought/traitsui#791
+            if is_current_backend_qt4():
+                with self.assertRaises(AssertionError):
+                    self.assertEqual(get_selected_indices(editor), [1, 2])
+                    self.assertEqual(editor.multi_selected_indices, [1, 2])
+                self.assertEqual(get_selected_indices(editor), [2, 0])
+                self.assertEqual(editor.multi_selected_indices, [0, 2])
+            else:
+                self.assertEqual(sorted(get_selected_indices(editor)), [1, 2])
+                self.assertEqual(sorted(editor.multi_selected_indices), [1, 2])
+
+    def test_list_str_editor_item_count(self):
+        gui = GUI()
+        model = ListStrModel()
+
+        # Without auto_add
+        with store_exceptions_on_all_threads(), \
+                create_ui(model, dict(view=get_view())) as ui:
+            gui.process_events()
+            editor = ui.get_editors("value")[0]
+            self.assertEqual(editor.item_count, 3)
+
+        # With auto_add
+        with store_exceptions_on_all_threads(), \
+                create_ui(model, dict(view=get_view(auto_add=True))) as ui:
+            gui.process_events()
+            editor = ui.get_editors("value")[0]
+            self.assertEqual(editor.item_count, 3)
+
+    def test_list_str_editor_refresh_editor(self):
+        # Smoke test for refresh_editor/refresh_
+        with store_exceptions_on_all_threads():
+            gui, editor = self.setup_gui(ListStrModel(), get_view())
+            if is_current_backend_qt4():
+                editor.refresh_editor()
+            elif is_current_backend_wx():
+                editor._refresh()
+            gui.process_events()
+
+    @skip_if_not_qt4
+    def test_list_str_editor_update_editor_single_qt(self):
+        # QT editor uses selected items as the source of truth when updating
+        model = ListStrModel()
+
+        with store_exceptions_on_all_threads():
+            gui, editor = self.setup_gui(model, get_view())
+
+            set_selected_single(editor, 0)
+            gui.process_events()
+            # Sanity check
+            self.assertEqual(editor.selected_index, 0)
+            self.assertEqual(editor.selected, "one")
+
+            model.value = ["two", "one"]
+            gui.process_events()
+
+            # Selected remains "one" and indices are updated accordingly
+            self.assertEqual(get_selected_indices(editor), [1])
+            self.assertEqual(editor.selected_index, 1)
+            self.assertEqual(editor.selected, "one")
+
+            # Removing "one" creates a case of no longer valid selection
+            model.value = ["two", "three"]
+            gui.process_events()
+
+            # Model selection is reset, but editor selection values are not
```

Comment:
If the view is given `selected` in the factory (see comment above), would the corresponding `selected` trait be unset, or be set to a wrong value?  If the selected trait is unset, then I think we are good as the model's selected trait matches the view. However the editor internal states being inconsistent might matter for subsequent change events. For example, if the user then selects index 1 again, would the value "three" be passed onto to the object's `selected` trait?  If it does, then perhaps everything is fine here. The last two assertions in this test would be temporary inconsistencies that don't actually matter.

Just wanted to check if this reveals a bug.

Reply:
I guess my comment is a bit confusing - the model here means the view model (what qt and wx use), not user `ListStrModel` defined in the test class. (Maybe I should change the name since `ListStrModel` is actually the same name qt model uses 😅)
`selected` trait on the user model is synced with `editor.selected`, so it would have the same value as in the test. I checked - if the user sets `selected_index=1`, nothing is updated because `selected_index` didn't really change and `_selected_index_change` isn't executed. However, if the user selects "three" in the UI, the update is initiated.

Reply:
I see. Looks like that is a bug. Let's open a separate issue for that if there isn't one already. Would you like to open it or should I?

Reply:
I'll open and reference the issue in the code 👍

Reply:
Thank you!

Reply:
Opened #872